File: test/language/array_types/cpp/ArraysMappingTest.cpp
#include "gtest/gtest.h"

#include "array_types/arrays_mapping/ArraysMapping.h"

#include "zserio/RebindAlloc.h"

namespace array_types
{
namespace arrays_mapping
{

using allocator_type = ArraysMapping::allocator_type;
using string_type = zserio::string<zserio::RebindAlloc<allocator_type, char>>;
template <typename T>
using vector_type = std::vector<T, zserio::RebindAlloc<allocator_type, T>>;

using BitBuffer = zserio::BasicBitBuffer<zserio::RebindAlloc<allocator_type, uint8_t>>;

class ArraysMappingTest : public ::testing::Test
{
protected:
    static const size_t FIXED_ARRAY_LENGTH = 5;
};

TEST_F(ArraysMappingTest, unsignedIntegerArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setUint8Array(vector_type<uint8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setUint16Array(vector_type<uint16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setUint32Array(vector_type<uint32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setUint64Array(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, signedIntegerArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setInt8Array(vector_type<int8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setInt16Array(vector_type<int16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setInt32Array(vector_type<int32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setInt64Array(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, unsignedBitfieldArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setBitfield8Array(vector_type<uint8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setBitfield16Array(vector_type<uint16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setBitfield32Array(vector_type<uint32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setBitfield63Array(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setUint8Value(8);
    arraysMapping.setVariableBitfieldLongArray(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableBitfieldIntArray(vector_type<uint32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableBitfieldShortArray(vector_type<uint16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableBitfieldByteArray(vector_type<uint8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setLength64(64);
    arraysMapping.setVariableBitfield64Array(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, signedBitfieldArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setIntfield8Array(vector_type<int8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setIntfield16Array(vector_type<int16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setIntfield32Array(vector_type<int32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setIntfield64Array(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setUint8Value(8);
    arraysMapping.setVariableIntfieldLongArray(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableIntfieldIntArray(vector_type<int32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableIntfieldShortArray(vector_type<int16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVariableIntfieldByteArray(vector_type<int8_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setLength32(64);
    arraysMapping.setVariableIntfield64Array(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, float16Array)
{
    ArraysMapping arraysMapping;
    arraysMapping.setFloat16Array(vector_type<float>(FIXED_ARRAY_LENGTH));
    arraysMapping.setFloat32Array(vector_type<float>(FIXED_ARRAY_LENGTH));
    arraysMapping.setFloat64Array(vector_type<double>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, variableUnsignedIntegerArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setVaruint16Array(vector_type<uint16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVaruint32Array(vector_type<uint32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVaruint64Array(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVaruintArray(vector_type<uint64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVarsizeArray(vector_type<uint32_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, variableSignedIntegerArrays)
{
    ArraysMapping arraysMapping;

    arraysMapping.setVarint16Array(vector_type<int16_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVarint32Array(vector_type<int32_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVarint64Array(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
    arraysMapping.setVarintArray(vector_type<int64_t>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, boolArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setBoolArray(vector_type<bool>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, stringArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setStringArray(vector_type<string_type>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, externArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setExternArray(vector_type<BitBuffer>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, compoundArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setCompoundArray(vector_type<TestStructure>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, enumArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setEnumArray(vector_type<TestEnum>(FIXED_ARRAY_LENGTH));
}

TEST_F(ArraysMappingTest, bitmaskArray)
{
    ArraysMapping arraysMapping;
    arraysMapping.setBitmaskArray(vector_type<TestBitmask>(FIXED_ARRAY_LENGTH));
}

} // namespace arrays_mapping
} // namespace array_types

File: test/extensions/language/array_types/java/array_types/ArraysMappingTest.java
package array_types;

import static org.junit.jupiter.api.Assertions.*;

import java.math.BigInteger;
import java.util.Arrays;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitBuffer;

import array_types.arrays_mapping.ArraysMapping;
import array_types.arrays_mapping.TestBitmask;
import array_types.arrays_mapping.TestEnum;
import array_types.arrays_mapping.TestStructure;

public class ArraysMappingTest
{
    @Test
    public void unsignedIntegerArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setUint8Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setUint16Array(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setUint32Array(new long[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final BigInteger[] bigIntegerArray = new BigInteger[FIXED_ARRAY_LENGTH];
        arraysMapping.setUint64Array(bigIntegerArray);
        assertTrue(Arrays.equals(bigIntegerArray, arraysMapping.getUint64Array()));
    }

    @Test
    public void signedIntegerArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setInt8Array(new byte[FIXED_ARRAY_LENGTH]);
        arraysMapping.setInt16Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setInt32Array(new int[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final long[] longArray = new long[FIXED_ARRAY_LENGTH];
        arraysMapping.setInt64Array(longArray);
        assertTrue(Arrays.equals(longArray, arraysMapping.getInt64Array()));
    }

    @Test
    public void unsignedBitfieldArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setBitfield8Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setBitfield16Array(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setBitfield32Array(new long[FIXED_ARRAY_LENGTH]);
        arraysMapping.setBitfield63Array(new long[FIXED_ARRAY_LENGTH]);
        arraysMapping.setUint8Value((short)8);
        arraysMapping.setVariableBitfieldLongArray(new BigInteger[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVariableBitfieldIntArray(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVariableBitfieldShortArray(new int[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final short[] unsignedByteArray = new short[FIXED_ARRAY_LENGTH];
        arraysMapping.setVariableBitfieldByteArray(unsignedByteArray);
        assertTrue(Arrays.equals(unsignedByteArray, arraysMapping.getVariableBitfieldByteArray()));

        arraysMapping.setLength64(BigInteger.valueOf(64));
        arraysMapping.setVariableBitfield64Array(new BigInteger[FIXED_ARRAY_LENGTH]);
    }

    @Test
    public void signedBitfieldArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setIntfield8Array(new byte[FIXED_ARRAY_LENGTH]);
        arraysMapping.setIntfield16Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setIntfield32Array(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setIntfield64Array(new long[FIXED_ARRAY_LENGTH]);
        arraysMapping.setUint8Value((short)8);
        arraysMapping.setVariableIntfieldLongArray(new long[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVariableIntfieldIntArray(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVariableIntfieldShortArray(new short[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final byte[] byteArray = new byte[FIXED_ARRAY_LENGTH];
        arraysMapping.setVariableIntfieldByteArray(byteArray);
        assertTrue(Arrays.equals(byteArray, arraysMapping.getVariableIntfieldByteArray()));

        arraysMapping.setLength32(64);
        arraysMapping.setVariableIntfield64Array(new long[FIXED_ARRAY_LENGTH]);
    }

    @Test
    public void floatArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setFloat16Array(new float[FIXED_ARRAY_LENGTH]);
        arraysMapping.setFloat32Array(new float[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final double[] float64Array = new double[FIXED_ARRAY_LENGTH];
        arraysMapping.setFloat64Array(float64Array);
        assertTrue(Arrays.equals(float64Array, arraysMapping.getFloat64Array()));
    }

    @Test
    public void variableUnsignedIntegerArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setVaruint16Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVaruint32Array(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVaruint64Array(new long[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVaruintArray(new BigInteger[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final int[] varSizeArray = new int[FIXED_ARRAY_LENGTH];
        arraysMapping.setVarsizeArray(varSizeArray);
        assertTrue(Arrays.equals(varSizeArray, arraysMapping.getVarsizeArray()));
    }

    @Test
    public void variableSignedIntegerArrays()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        arraysMapping.setVarint16Array(new short[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVarint32Array(new int[FIXED_ARRAY_LENGTH]);
        arraysMapping.setVarint64Array(new long[FIXED_ARRAY_LENGTH]);

        // just do something with arraysMapping not to have SpotBugs warning
        final long[] varIntArray = new long[FIXED_ARRAY_LENGTH];
        arraysMapping.setVarintArray(varIntArray);
        assertTrue(Arrays.equals(varIntArray, arraysMapping.getVarintArray()));
    }

    @Test
    public void boolArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        // just do something with arraysMapping not to have SpotBugs warning
        final boolean[] boolArray = new boolean[FIXED_ARRAY_LENGTH];
        arraysMapping.setBoolArray(boolArray);
        assertTrue(Arrays.equals(boolArray, arraysMapping.getBoolArray()));
    }

    @Test
    public void stringArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();
        final String[] stringArray = new String[FIXED_ARRAY_LENGTH];
        arraysMapping.setStringArray(stringArray);
        assertTrue(Arrays.equals(stringArray, arraysMapping.getStringArray()));
    }

    @Test
    public void bytesArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();
        final byte[][] bytesArray = new byte[FIXED_ARRAY_LENGTH][];
        arraysMapping.setBytesArray(bytesArray);
        assertTrue(Arrays.equals(bytesArray, arraysMapping.getBytesArray()));
    }

    @Test
    public void externArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();
        final BitBuffer[] externArray = new BitBuffer[FIXED_ARRAY_LENGTH];
        arraysMapping.setExternArray(externArray);
        assertTrue(Arrays.equals(externArray, arraysMapping.getExternArray()));
    }

    @Test
    public void compoundArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        // just do something with arraysMapping not to have SpotBugs warning
        final TestStructure[] objectArray = new TestStructure[FIXED_ARRAY_LENGTH];
        arraysMapping.setCompoundArray(objectArray);
        assertTrue(Arrays.equals(objectArray, arraysMapping.getCompoundArray()));
    }

    @Test
    public void enumArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        // just do something with arraysMapping not to have SpotBugs warning
        final TestEnum[] enumArray = new TestEnum[FIXED_ARRAY_LENGTH];
        arraysMapping.setEnumArray(enumArray);
        assertTrue(Arrays.equals(enumArray, arraysMapping.getEnumArray()));
    }

    @Test
    public void bitmaskArray()
    {
        ArraysMapping arraysMapping = new ArraysMapping();

        // just do something with arraysMapping not to have FindBugs warning
        final TestBitmask[] bitmaskArray = new TestBitmask[FIXED_ARRAY_LENGTH];
        arraysMapping.setBitmaskArray(bitmaskArray);
        assertTrue(Arrays.equals(bitmaskArray, arraysMapping.getBitmaskArray()));
    }

    private static final int FIXED_ARRAY_LENGTH = 5;
}
